operator/scheduler/client.go

```diff
@@ -30,40 +32,129 @@ import (
+	mu               sync.Mutex
 }
 
+//  connect on demand by add getConnection(namespace) which if not existing calls connect to sheduler.
+// For this will need to know ports (hardwire for now to 9004 and 9044 - ssl comes fom envvar - so always
+// the same for all schedulers
+
 func NewSchedulerClient(logger logr.Logger, client client.Client, recorder record.EventRecorder) *SchedulerClient {
 	opts := []grpc.CallOption{
 		grpc.MaxCallSendMsgSize(math.MaxInt32),
 		grpc.MaxCallRecvMsgSize(math.MaxInt32),
 	}
 
 	return &SchedulerClient{
-		Client:      client,
-		logger:      logger.WithName("schedulerClient"),
-		callOptions: opts,
-		recorder:    recorder,
+		Client:         client,
+		logger:         logger.WithName("schedulerClient"),
+		callOptions:    opts,
+		recorder:       recorder,
+		seldonRuntimes: make(map[string]*grpc.ClientConn),
+	}
+}
+
+func getSchedulerHost(namespace string) string {
+	return fmt.Sprintf("seldon-scheduler.%s", namespace)
+}
+
+func (s *SchedulerClient) startEventHanders(namespace string, conn *grpc.ClientConn) {
+	// Subscribe the event streams from scheduler
+	go func() {
+		err := s.SubscribeModelEvents(context.Background(), namespace, conn)
+		if err != nil {
+			s.RemoveConnection(namespace)
+		}
+	}()
+	go func() {
+		err := s.SubscribeServerEvents(context.Background(), namespace, conn)
+		if err != nil {
+			s.RemoveConnection(namespace)
+		}
+	}()
+	go func() {
+		err := s.SubscribePipelineEvents(context.Background(), namespace, conn)
+		if err != nil {
+			s.RemoveConnection(namespace)
+		}
+	}()
+	go func() {
+		err := s.SubscribeExperimentEvents(context.Background(), namespace, conn)
+		if err != nil {
+			s.RemoveConnection(namespace)
+		}
+	}()
+}
+
+func (s *SchedulerClient) RemoveConnection(namespace string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if conn, ok := s.seldonRuntimes[namespace]; ok {
+		delete(s.seldonRuntimes, namespace)
+		err := conn.Close()
+		if err != nil {
+			s.logger.Error(err, "Failed to close grpc connection to scheduler", "namespace", namespace)
+		}
 	}
 }
 
-func (s *SchedulerClient) ConnectToScheduler(host string, plainTxtPort int, tlsPort int) error {
+func (s *SchedulerClient) smokeTestConnection(conn *grpc.ClientConn) error {
+	grcpClient := scheduler.NewSchedulerClient(conn)
+
+	stream, err := grcpClient.SubscribeModelStatus(context.TODO(), &scheduler.ModelSubscriptionRequest{SubscriberName: "seldon manager"}, grpc_retry.WithMax(1))
+	if err != nil {
+		return err
+	}
+	err = stream.CloseSend()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *SchedulerClient) getConnection(namespace string) (*grpc.ClientConn, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if conn, ok := s.seldonRuntimes[namespace]; !ok {
+		var err error
+		conn, err = s.connectToScheduler(getSchedulerHost(namespace), namespace, 9004, 9044)
+		if err != nil {
+			return nil, err
+		}
+		err = s.smokeTestConnection(conn)
```

Comment:
perhaps add a note to describe why we do a smoke test to scheduler? Make sure it is up before proceeding further?